### ISSUE TITLE
macOS chart focus

### ISSUE THREAD

**Issue by jlixfeld** [louisnw01/lightweight-charts-python#55]
I'm having issues getting the chart to take focus for keystrokes in macOS with this example from the docs:

```pyton
import asyncio
import pandas as pd

from lightweight_charts import Chart


def get_bar_data(symbol, timeframe):
    if symbol not in ('AAPL', 'GOOGL', 'TSLA'):
        print(f'No data for "{symbol}"')
        return pd.DataFrame()
    return pd.read_csv(f'tmp/lightweight-charts-python/examples/6_callbacks/bar_data/{symbol}_{timeframe}.csv')


class API:
    def __init__(self):
        self.chart = None  # Changes after each callback.

    async def on_search(self, searched_string):  # Called when the user searches.
        new_data = get_bar_data(searched_string, self.chart.topbar['timeframe'].value)
        if new_data.empty:
            return
        self.chart.topbar['corner'].set(searched_string)
        self.chart.set(new_data)

    async def on_timeframe_selection(self):  # Called when the user changes the timeframe.
        new_data = get_bar_data(self.chart.topbar['corner'].value, self.chart.topbar['timeframe'].value)
        if new_data.empty:
            return
        self.chart.set(new_data)
            
    async def on_horizontal_line_move(self, line_id, price):
        print(f'Horizontal line moved to: {price}')


async def main():
    api = API()

    chart = Chart(api=api, topbar=True, searchbox=True, toolbox=True)
    chart.legend(True)

    chart.topbar.textbox('corner', 'TSLA')
    chart.topbar.switcher('timeframe', api.on_timeframe_selection, '1min', '5min', '30min', default='5min')

    df = get_bar_data('TSLA', '5min')
    chart.set(df)
    
    chart.horizontal_line(200, interactive=True)

    await chart.show_async(block=True)


if __name__ == '__main__':
    asyncio.run(main())
```

I can click through the timeframes, and the mouse moves the crosshairs, etc, but I can't type to activate the search bar (assuming it works the same way that TradingView does where you type letters it defaults to symbols and numbers defaults to timeframes).

Is there some additional magic required in macOS for focus?

Same issue whether I run the code for iTerm,app, Terminal.app, or VS Code terminal window.

**Comment by louisnw01** [comment]
Hey @jlixfeld,

At the moment the search detection does have some issues with detection but it should still work; when you run the program, try leaving your mouse cursor out of the chart window when it initially opens, and only once its loaded move your cursor into the chart window.

The library should detect your mouse entering and allow you to search. Let me know if this works for the time being!

**Comment by jlixfeld** [comment]
Same issue.  The following capture shows initial load with keystroke, mouseover with keystroke.  click-to-highlight with keystroke, and finally clicking symbol with keystroke.


https://github.com/louisnw01/lightweight-charts-python/assets/17339017/22a39716-507f-4126-9b80-be76dc7f2f03

**Comment by louisnw01** [comment]
Unable to reproduce on Monterey.

Could you use `Chart(debug=True)`, right click the window to open the web inspector and let me know if theres any errors showing in the web console?

**Comment by jlixfeld** [comment]
I'm on Ventura 13.4.1 (c)

There was an error generated when I click on the chart after moving it:

<img width="1301" alt="Screenshot 2023-08-01 at 3 36 48 PM" src="https://github.com/louisnw01/lightweight-charts-python/assets/17339017/cba452c7-7f8c-43ce-8a60-555558e9b843">

Otherwise, nothing else logged.

**Comment by louisnw01** [comment]
The latest version removes that error, however I believe it is unrelated.

Could you try adding this to your code:
```python
chart.run_script(f'{chart.id}.wrapper.dispatchEvent(new Event("mouseover"))')
```

**Comment by jlixfeld** [comment]
Still not able to access the searcher on 1.0.15 with that code snippet.

However, I'll point out an additional observation in case it's relevant:

After I've loaded a chart, I can right click on the chart and have access to the Inspect Element option.  If I add a line to the chart, I am able to right click on the line to delete it, or right click anywhere else on the chart to access the Inspect Element option.  If I delete the line I just added, no matter where I click on the chart,I no longer have access to the Inspect Element option. I only have access to the Delete Line option, even though there are no lines on the chart.  If I go ahead and add a line again, then I have access to the inspect element option again.

**Comment by louisnw01** [comment]
> However, I'll point out an additional observation in case it's relevant:
> 
> After I've loaded a chart, I can right click on the chart and have access to the Inspect Element option. If I add a line to the chart, I am able to right click on the line to delete it, or right click anywhere else on the chart to access the Inspect Element option. If I delete the line I just added, no matter where I click on the chart,I no longer have access to the Inspect Element option. I only have access to the Delete Line option, even though there are no lines on the chart. If I go ahead and add a line again, then I have access to the inspect element option again.

This appears to be a seperate bug; I'll fix in the next update.

As for the searcher issue, could you try running this code? As you mouse over the box it should change color, and when you click the button the box should also change color.

```python
import webview


html = f'''
<body>
<script>
let box = document.createElement('div')
box.style.width = '50%'
box.style.height = '50%'
box.style.backgroundColor = 'blue'
let button = document.createElement('button')
button.innerText = 'red'
button.onclick = () => box.dispatchEvent(new Event("mouseover"))
document.body.appendChild(box)
document.body.appendChild(button)
box.addEventListener('mouseover', (event) => box.style.backgroundColor = 'red')
box.addEventListener('mouseout', (event) => box.style.backgroundColor = 'blue')
</script>
</body>
'''

webview.create_window('', html=html)
webview.start(debug=True)
```

**Comment by jlixfeld** [comment]
> As for the searcher issue, could you try running this code? As you mouse over the box it should change color, and when you click the button the box should also change color.

Ack.  No issues there.

At the risk of sounding completely obtuse, how is the searcher supposed to be activated?  Automagically detects keystrokes when in focus a'la Trading View's app?  Or, hover over the symbol or clicking the symbol or key sequence or something else?

**Comment by louisnw01** [comment]
Yes pretty much; as there can be multiple charts in the same window (subcharts) it keeps track of the latest chart to be moused over, and opens the search box for that pane when any key is pressed.

Strange that works, I've never had anyone encounter this issue before.

And to confirm, this causes the same issue:

```python
import pandas as pd
from lightweight_charts import Chart

if __name__ == '__main__':
    chart = Chart(topbar=True, searchbox=True)
    chart.set(pd.read_csv('ohlcv.csv'))
    chart.show(block=True)

```

My instincts tell me this is a system issue, but the fact that the color box script worked means it may not be.

**Comment by jlixfeld** [comment]
> And to confirm, this causes the same issue:

Yup.

https://github.com/louisnw01/lightweight-charts-python/assets/17339017/60da6c80-4907-4030-af68-b5e82cd91938

**Comment by louisnw01** [comment]
> > And to confirm, this causes the same issue:
> 
> Yup.
> 
>  Screen.Recording.2023-08-02.at.7.01.38.PM.mov

When you type in that video, it looks like the focus is being taken from the Chart window into your IDE's console.

See if there's a setting to disable that and it should work.

Louis

**Comment by jlixfeld** [comment]
Would you lean the same way if I said that the same thing happens if I do it in any terminal; built in Terminal.app, iTerm and this IDE? :)

**Comment by louisnw01** [comment]
> Would you lean the same way if I said that the same thing happens if I do it in any terminal; built in Terminal.app, iTerm and this IDE? :)

Maybe not; Could you send a video of you running the same code in terminal?

**Comment by jlixfeld** [comment]
Here ya go.

https://github.com/louisnw01/lightweight-charts-python/assets/17339017/74f52175-359b-4a73-bfcd-b9b15b204b72

**Comment by jlixfeld** [comment]
Here you go (take two)

https://github.com/louisnw01/lightweight-charts-python/assets/17339017/1184a5cf-7eeb-4f41-a0a7-d0c681bc324f

**Comment by louisnw01** [comment]
Hmm, that also changes focus. It appears to be an issue at a lower level than the library; something is causing focus to jump away from the python window and back to the console before the key is detected in Javascript.

If it were an issue with the library, I wouldn't expect that behaviour. If you can't find a fix by tomorrow I'll write another pywebview script to confirm.

Louis

**Comment by louisnw01** [comment]
Try this code, when you press a key the box should change color:

```python
import webview


html = f'''
<body>
<script>
let box = document.createElement('div')
box.style.width = '50%'
box.style.height = '50%'
box.style.backgroundColor = 'blue'
let button = document.createElement('button')
button.innerText = 'blue'
button.onclick = () => box.style.backgroundColor = 'blue'
document.body.appendChild(box)
document.body.appendChild(button)
document.body.addEventListener('keydown', (event) => {{
event.preventDefault()
box.style.backgroundColor = 'red'
}})
</script>
</body>
'''

webview.create_window('', html=html)
webview.start(debug=True)
```

**Comment by jlixfeld** [comment]
Negative.

Last night I asked ChatGPT for an MRE to file https://github.com/r0x0r/pywebview/issues/1187 and they couldn't replicate it.

**Comment by louisnw01** [comment]
> Negative. Keystrokes in the terminal, but no color change.
> 
> Also, last night I asked ChatGPT for an MRE to file [r0x0r/pywebview#1187](https://github.com/r0x0r/pywebview/issues/1187) and they couldn't replicate it. However, in their example, text was on the terminal as well as in the text window. That surprised me as I wasn't expecting it to appear in the terminal in a working case, unless having debug enabled makes that happen?

Just had a look; the MRE you gave fires a callback to Python and prints text typed in the JS textbox in the console, so it looks similar when it works vs when it doesn't.

I think your best bet is to look for other people experiencing this issue not just in pywebview but other libraries, as it seems to be related to your system rather than pywebview. I found [this](https://github.com/pygame/pygame/issues/203) which seems like the same thing.

**Comment by jlixfeld** [comment]
Circling back for anybody else who finds this:

The issue in my case was that I was using `python3` installed by `pyenv`.  [This](https://stackoverflow.com/a/4781033/4907339) post tipped me off.

Here are the steps for how I fixed it in my particular environment:
- Uninstalled all versions of python installed by `pyenv`, and then `pyenv` itself.
- Installed `python3` via `homebrew` (which at the time gave me `python 3.11.5`): `brew install python@3.11`
- Getting rid of `pyenv` broke my `poetry` install, so I had to uninstall/re-install it:
    - Trying the official poetry installer, I hit [this](https://stackoverflow.com/questions/70883863/poetry-install-fails-on-macos-11-6-xcode-13-dyld-library-not-loaded-executab) again so I instead installed poetry via `pipx` after installing `pipx` via `homebrew` as well: `brew install pipx`, `pipx install poetry`
- Ran `poetry shell` to re-create the virtual environment, then `poetry install` to install my project dependencies.
- Ran [this](https://lightweight-charts-python.readthedocs.io/en/latest/examples/events.html#topbar-example) example, and it finally worked perfectly.
